From 232d1901c6c726ef1a589c27093db00b3a5ffff6 Mon Sep 17 00:00:00 2001
From: Baptiste Chauloux <baptiste.chauloux@st.com>
Date: Mon, 30 Jan 2017 09:29:00 +0100
Subject: [PATCH] Disable systemd services start Done by udev

Signed-off-by: Baptiste Chauloux <baptiste.chauloux@st.com>
---
 alsactl/Makefile.am | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/alsactl/Makefile.am b/alsactl/Makefile.am
index ff865dc..e565cc4 100644
--- a/alsactl/Makefile.am
+++ b/alsactl/Makefile.am
@@ -31,9 +31,7 @@ install-data-hook:
 	$(MKDIR_P) -m 0755 \
 		$(DESTDIR)$(systemdsystemunitdir)/basic.target.wants
 	( cd $(DESTDIR)$(systemdsystemunitdir)/basic.target.wants && \
-		rm -f alsa-state.service alsa-restore.service && \
-		$(LN_S) ../alsa-state.service alsa-state.service && \
-		$(LN_S) ../alsa-restore.service alsa-restore.service)
+		rm -f alsa-state.service alsa-restore.service)
 
 endif
 
-- 
2.7.4

