From c7fa0a47956570682e289937bf6f2d0516f491c6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andr=C3=A9=20Draszik?= <adraszik@tycoint.com>
Date: Thu, 23 Mar 2017 12:24:02 +0000
Subject: [PATCH 2/2] build: don't need charset.alias when building for musl
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Upstream-Status: Inappropriate [required for m4 1.4.9 (GPLv2) recipe only]
Signed-off-by: André Draszik <adraszik@tycoint.com>
---
 lib/Makefile.am | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/lib/Makefile.am b/lib/Makefile.am
index 2575743..4dc17fe 100644
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -290,7 +290,11 @@ all-local: charset.alias ref-add.sed ref-del.sed
 charset_alias = $(DESTDIR)$(libdir)/charset.alias
 charset_tmp = $(DESTDIR)$(libdir)/charset.tmp
 install-exec-local: all-local
-	test $(GLIBC21) != no || $(mkinstalldirs) $(DESTDIR)$(libdir)
+	case '$(host_os)' in \
+	    linux-musl*) \
+	      : ;;\
+	    *) \
+	test $(GLIBC21) != no || $(mkinstalldirs) $(DESTDIR)$(libdir) ;\
 	if test -f $(charset_alias); then \
 	  sed -f ref-add.sed $(charset_alias) > $(charset_tmp) ; \
 	  $(INSTALL_DATA) $(charset_tmp) $(charset_alias) ; \
@@ -301,7 +305,8 @@ install-exec-local: all-local
 	    $(INSTALL_DATA) $(charset_tmp) $(charset_alias) ; \
 	    rm -f $(charset_tmp) ; \
 	  fi ; \
-	fi
+	fi ;\
+	esac
 
 uninstall-local: all-local
 	if test -f $(charset_alias); then \
-- 
2.11.0

