From 1ab83a81d6b7353093f8daf9b30b11a6db559691 Mon Sep 17 00:00:00 2001
From: Mustapha Ghanmi <mustapha.ghanmi@st.com>
Date: Tue, 31 Jul 2018 14:30:55 +0200
Subject: [PATCH] waylandsink: add support of UYVY color format

Signed-off-by: Mustapha Ghanmi <mustapha.ghanmi@st.com>
---
 ext/wayland/waylandpool.c   | 1 +
 ext/wayland/wlvideoformat.c | 1 +
 2 files changed, 2 insertions(+)

diff --git a/ext/wayland/waylandpool.c b/ext/wayland/waylandpool.c
index 663a627..7326095 100644
--- a/ext/wayland/waylandpool.c
+++ b/ext/wayland/waylandpool.c
@@ -437,6 +437,7 @@ get_format_bpp (int format)
     case DRM_FORMAT_ARGB8888:
       return 32;
     case DRM_FORMAT_RGB565:
+    case DRM_FORMAT_UYVY:
       return 16;
     default:
       GST_WARNING ("Unknown format: %d", format);
diff --git a/ext/wayland/wlvideoformat.c b/ext/wayland/wlvideoformat.c
index 40683eb..dd38c28 100644
--- a/ext/wayland/wlvideoformat.c
+++ b/ext/wayland/wlvideoformat.c
@@ -113,6 +113,7 @@ static const wl_DmabufVideoFormat dmabuf_formats[] = {
   {DRM_FORMAT_YUYV, GST_VIDEO_FORMAT_YUY2},
   {DRM_FORMAT_NV12, GST_VIDEO_FORMAT_NV12},
   {DRM_FORMAT_YUYV, GST_VIDEO_FORMAT_YUY2},
+  {DRM_FORMAT_UYVY, GST_VIDEO_FORMAT_UYVY},
   {DRM_FORMAT_YUV420, GST_VIDEO_FORMAT_I420},
 };
 
-- 
2.7.4

